### PR TITLE
Mailbox 268 2

### PR DESCRIPTION
**Pull request by rouazana**
Write AttachmentManager class. Call AttachmentManager.appendAttachment from StoreMessageManager.appendMessageToStore for every Attachment found with the attachment finder and then store the Message.

It should only handle first level attachment. In case of parsing error, store the message anyway.

Test cases :

    save mails, retrieve them and load attachments. 1.1 when no attachment 1.2 with 1 attachment 1.3 with several attachments 1.4 with embedded messages
    deliver the same emails with attachments to several users

Don't mock MessageManager and AttachmentManager. Write a default dummy AttachManager for test Take care of spring !

### CODE REVIEW

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerAttachmentTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    private CassandraMailboxManager mailboxManager;
+    private CassandraMailboxManager parseFailingMailboxManager;
+
+    public CassandraMailboxManagerAttachmentTest() throws Exception {
+        mailboxSessionMapperFactory = new CassandraMailboxSessionMapperFactory(
+                new CassandraUidProvider(cassandra.getConf()),
+                new CassandraModSeqProvider(cassandra.getConf()),
+                cassandra.getConf(),
+                cassandra.getTypesProvider());
+        mailboxManager = new CassandraMailboxManager(mailboxSessionMapperFactory, null, new JVMMailboxPathLocker(), new MessageParser());
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerAttachmentTest.java]: could you give a name to `null` ?

Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerAttachmentTest.java]: we should not use a JVM locker, don't we ?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerTest.java [hunk 2]
```diff
@@ -20,8 +20,8 @@
 
 import org.apache.james.mailbox.store.mail.model.AbstractMailboxMapperTest;
 
-public class CassandraMailboxMapperTest extends AbstractMailboxMapperTest {
-    public CassandraMailboxMapperTest() {
+public class CassandraMailboxManagerTest extends AbstractMailboxMapperTest {
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerTest.java]: Manager extends Mapper ?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMailboxManagerAttachmentTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+public class InMemoryMailboxManagerAttachmentTest extends AbstractMailboxManagerAttachmentTest {
+
+    private InMemoryMailboxSessionMapperFactory mailboxSessionMapperFactory;
+    private InMemoryMailboxManager mailboxManager;
+    private InMemoryMailboxManager parseFailingMailboxManager;
+
+    public InMemoryMailboxManagerAttachmentTest() throws Exception {
+        mailboxSessionMapperFactory = new InMemoryMailboxSessionMapperFactory();
+        mailboxManager = new InMemoryMailboxManager(mailboxSessionMapperFactory, null, new JVMMailboxPathLocker(), new UnionMailboxACLResolver(), null, new MessageParser());
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMailboxManagerAttachmentTest.java]: could you name `null` ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 4]
```diff
@@ -409,6 +423,15 @@ public Long execute() throws MailboxException {
 
     }
 
+    private List<Attachment> parseAttachments(SharedFileInputStream contentIn) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: I don't think it parses attachment, it parses an email and extract some Attachment

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 5]
```diff
@@ -607,12 +635,16 @@ private void trimFlags(Flags flags, MailboxSession session) {
...
-        final MessageMapper mapper = mapperFactory.getMessageMapper(session);
+    protected MessageMetaData appendMessageToStore(final MailboxMessage message, final List<Attachment> attachments, MailboxSession session) throws MailboxException {
+        final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+        final AttachmentMapper attachmentMapper = mapperFactory.getAttachmentMapper(session);
         return mapperFactory.getMessageMapper(session).execute(new Mapper.Transaction<MessageMetaData>() {
 
             public MessageMetaData run() throws MailboxException {
-                return mapper.add(getMailboxEntity(), message);
+                for (Attachment attachment: attachments) {
+                    attachmentMapper.storeAttachment(attachment);
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: we should pass the list to store to let the backend use async apis, don't you think ?

Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: what about error during storage ?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerAttachmentTest.java (hunk 1 above)
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerAttachmentTest.java]: NoMailboxPathLocker is here for that.

**Comment by rouazana** [comment]
test this please

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerTest.java (hunk 2 above)
Comment (aduprat) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerTest.java]: Manapper ?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerAttachmentTest.java (hunk 1 above)
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerAttachmentTest.java]: ok for the 3 comments

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerTest.java (hunk 2 above)
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxManagerTest.java]: No idea where this change comes from. Fixed.

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMailboxManagerAttachmentTest.java (hunk 3 above)
Comment (rouazana) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/mail/InMemoryMailboxManagerAttachmentTest.java]: ok

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 4 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: ok

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 5 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: > what about error during storage ?
It should throw, and this seems the correct behaviour

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMailboxManagerTest.java [hunk 6]
```diff
@@ -92,7 +95,7 @@ protected void createMailboxManager() throws MailboxException {
             CASSANDRA.getConf(),
             CASSANDRA.getTypesProvider());
 
-        final CassandraMailboxManager manager = new CassandraMailboxManager(mapperFactory, null, new JVMMailboxPathLocker());
+        final CassandraMailboxManager manager = new CassandraMailboxManager(mapperFactory, null, new JVMMailboxPathLocker(), new MessageParser());
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraMailboxManagerTest.java]: Not your commit but should be NoMailboxPathLocker IMO

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMailboxManagerAttachmentTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+        Iterator<MailboxMessage> messages = messageMapper.findInMailbox(inbox, MessageRange.all(), FetchType.Full, 1);
+        Iterator<MailboxMessage> user2Messages = user2MessageMapper.findInMailbox(user2Inbox, MessageRange.all(), FetchType.Full, 1);
+        assertThat(messages.hasNext()).isTrue();
+        List<AttachmentId> attachmentsIds = messages.next().getAttachmentsIds();
+        assertThat(attachmentsIds).hasSize(1);
+        assertThat(user2Messages.hasNext()).isTrue();
+        List<AttachmentId> user2AttachmentsIds = user2Messages.next().getAttachmentsIds();
+        assertThat(attachmentsIds.equals(user2AttachmentsIds)).isTrue();
+    }
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMailboxManagerAttachmentTest.java]: Could you add some tests with contentType != Full ?

Just to make sure attachements are not retrieved....

Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMailboxManagerAttachmentTest.java]: Already done here: https://github.com/linagora/james-project/pull/247/commits/f53d0eb93045212d778b211384dde48ef1cefc99#diff-18981fdc6e5a270d2de4ef7632e58f4aR285

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 5 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: > we should pass the list to store to let the backend use async apis, don't you think ?

done in a other commit

**Comment by chibenwa** [comment]
:+1:

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentMapper.java [hunk 8]
```diff
@@ -95,15 +99,34 @@ private Attachment attachment(Row row) {
...
+        return cassandraAsyncExecutor.executeVoid(
+                insert.bind()
+                    .setString(ID, attachment.getAttachmentId().getId())
+                    .setBytes(PAYLOAD, ByteBuffer.wrap(IOUtils.toByteArray(attachment.getStream())))
+                    .setString(TYPE, attachment.getType())
+                    .setLong(SIZE, attachment.getSize())
+            );
+    }
+
+    static class RuntimeIOException extends RuntimeException {};
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentMapper.java]: unused class

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AbstractAttachmentMapperTest.java [hunk 9]
```diff
@@ -69,4 +71,20 @@ public void getAttachmentShouldReturnTheAttachmentWhenReferenced() throws Except
...
     }
+
+    @Test
+    public void getAttachmentShouldReturnTheAttachmentsWhenMultipleStored() throws Exception {
+        //Given
+        Attachment expected1 = Attachment.from("payload1".getBytes(), "content1");
+        Attachment expected2 = Attachment.from("payload2".getBytes(), "content2");
+        AttachmentId attachmentId1 = expected1.getAttachmentId();
+        AttachmentId attachmentId2 = expected2.getAttachmentId();
+        attachmentMapper.storeAttachments(ImmutableList.of(expected1, expected2));
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AbstractAttachmentMapperTest.java]: I think the when is in fact this line

**Comment by mbaechler** [comment]
:+1:

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentMapper.java (hunk 8 above)
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentMapper.java]: oops :)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AbstractAttachmentMapperTest.java (hunk 9 above)
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AbstractAttachmentMapperTest.java]: Yep, comments sux.

**Comment by aduprat** [comment]
:+1:

**Pull request merged**